### PR TITLE
[feat] Impl docker containerization

### PR DESCRIPTION
**Pull request by EpicPi**
## Changes made
Backend and mongo are now run through docker. 

New scripts system in place - `bin/dev`

Basic usage is 
1. `bin/dev setup` (to undo after returning to non-docker branch, `pre-commit install`)
2. `bin/dev build`
3. `bin/dev up`
4. `bin/dev debug` to run backend
5. `bin/dev webpack` to run fronend
6. `bin/dev pytest args` to run tests
7. `bin/dev pre-commit args` to run pre-commit
8. `bin/dev down` to turn everything off

N.b. 
New dev.env file must be used - located under "New Dot Env" in notion.
Must install docker: https://docs.docker.com/get-docker/

## Work left to be done
- frontend is not run in containers yet - will be tackled in a follow up pr

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+ENV PIPENV_VENV_IN_PROJECT 1
+ENV FLASK_APP 'server:app'
+
+# cli args to copy over host machine user
+ARG user_id
+ARG group_id
+
+# give evrything to the worker
```
Comment (SammyBaek) [inline Dockerfile:17]:
```suggestion
# give everything to the worker
```
[nit] spelling

File: client/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+FROM node:latest
```
Comment (SammyBaek) [inline client/Dockerfile:1]: Didn't you mention this was not supposed to be here? Or are we keeping it for the (near) future possibility of containerizing everything?

**Review by SammyBaek — APPROVED** [review]
Looks great! Please wait for our local test to finish to merge.

File: bin/dev [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+# Run js tests
+runJstest(){
+    cd client
+    npm run test $*
+}
+
+# Run precommit inside container
+runPrecommit(){
+    findBackendContainerId container_id
+    cmd="docker exec -w /home/worker/app/ "$container_id" pre-commit "$*
```
Comment (SammyBaek) [inline bin/dev:85]:
```suggestion
    cmd="docker exec -t -w /home/worker/app/ "$container_id" pre-commit "$*
```

runs pseudo-TTY, it allows colors
https://docs.docker.com/engine/reference/commandline/exec/#options

image

File: bin/dev [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+# Run frontend webpack and attach shell
+runWebpack(){
+    cd client
+    npm run start
+}
+
+# Run pytest
+runPytest(){
+    findBackendContainerId container_id
+    cmd="docker exec "$container_id" pytest "$*
```
Comment (SammyBaek) [inline bin/dev:71]:
```suggestion
    cmd="docker exec -t "$container_id" pytest "$*
```
runs pseudo-TTY, it allows colors
https://docs.docker.com/engine/reference/commandline/exec/#options

image

File: client/Dockerfile (hunk 2 above)
Comment (EpicPi) [reply client/Dockerfile:1]: keeping for future work.

**Pull request merged**